### PR TITLE
feat: RPC - add `starknet_getClassHashAt`

### LINKED ISSUE #196: feat: RPC - starknet_getClassHashAt
Get the contract class hash in the given block for the contract deployed at the given address

### PR DESCRIPTION
**Pull request by ftupas**
Implement `starknet_getClassHashAt` RPC that gets the contract class hash in the given block for the contract deployed at the given address

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: #196 

# What is the new behavior?
- Add `getClassHashAt` to RPC client
- Test `getClassHashAt` RPC method
- Fix #196 

image

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

# Other information

### CODE REVIEW

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -196,6 +197,62 @@ where
...
+        let class_hash = self
+            .overrides
+            .for_block_hash(self.client.as_ref(), substrate_block_hash)
+            .contract_class_hash_by_address(substrate_block_hash, contract_address_wrapped)
+            .ok_or_else(|| {
+                error!("Failed to retrieve contract class hash at '{contract_address}'");
+                StarknetRpcApiError::ContractNotFound
+            })?;
+
+        Ok(StarkFelt::new(class_hash).unwrap().to_string())
```
Comment (EvolveArt) [inline crates/client/rpc/src/lib.rs:254]: no `unwrap` please

File: crates/client/storage/src/overrides/mod.rs [hunk 2]
```diff
@@ -85,4 +96,12 @@ where
 
         api.current_block(block_hash).ok()
     }
+    fn contract_class_hash_by_address(
```
Comment (EvolveArt) [inline crates/client/storage/src/overrides/mod.rs:99]: comment + formatting

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left few comments, please improve integration tests (add more cases and test return values)

**Comment by ftupas** [comment]
> left few comments, please improve integration tests (add more cases and test return values)

Thanks for the review @EvolveArt , will update accordingly ✌️

File: crates/client/storage/src/overrides/mod.rs (hunk 2 above)
Comment (ftupas) [reply crates/client/storage/src/overrides/mod.rs:127]: Updated with comment and formatting 👍

File: crates/client/rpc-core/src/lib.rs [hunk 3]
```diff
@@ -11,6 +11,7 @@ use serde::{Deserialize, Serialize};
 pub type FieldElement = String;
 pub type BlockNumber = u64;
 pub type BlockHash = FieldElement;
+pub type Address = String;
```
[the remark this answers is not in the document]
Comment (ftupas) [reply crates/client/rpc-core/src/lib.rs]: Decided to use the existing `ContractAddress` instead

File: crates/client/rpc/src/lib.rs (hunk 1 above)
Comment (ftupas) [reply crates/client/rpc/src/lib.rs]: Update this one 👍

**Comment by abdelhamidbakhta** [comment]
hey @ftupas what is the status of this PR ?

**Comment by ftupas** [comment]
> hey @ftupas what is the status of this PR ?

Hi @abdelhamidbakhta, it's up for review normally I've resolved the conflicts 🤙 I've also improved the integration test by adding more test cases

**Comment by abdelhamidbakhta** [comment]
> > hey @ftupas what is the status of this PR ?
> 
> Hi @abdelhamidbakhta, it's up for review normally I've resolved the conflicts 🤙 I've also improved the integration test by adding more test cases

awesome thanks for the prompt reply. I will leave @EvolveArt continue the review then

**Review by EvolveArt — APPROVED** [review]
lgtm

**Pull request merged**